### PR TITLE
Provide MPI information for KNL and some new GCC compile flags.

### PR DESCRIPTION
**Pull request by KineticTheory**
+ For KNL, set
```
  MPI_CORES_PER_CPU = 17
  MPI_PHYSICAL_CORES = 4
  MPIEXEC_MAX_NUMPROCS 68
```
+ For g++, test if option -Wunused-macros is available.
  - If so, use it.
  - Clean up some code that had unused CPP macros.
* [Pre-Merge Code Review](https://github.com/lanl/Draco/wiki/Style-Guide)
  * [x] Travis CI checks pass
  * [x] Code coverage does not decrease
  * [x] [Valgrind test passes](https://rtt.lanl.gov/CDash/index.php?project=Draco)
  * [x] [Toss2 checks pass](https://rtt.lanl.gov/CDash/index.php?project=Draco)
  * [x] [Toss3 checks pass](https://rtt.lanl.gov/CDash/index.php?project=Draco)
  * [x] [Trinitite checks pass](https://rtt.lanl.gov/CDash/index.php?project=Draco)
  * [ ] Code reviewed/approved, sufficient DbC checks, testing, documentation

### CODE REVIEW

File: config/unix-g++.cmake [hunk 1]
```diff
@@ -143,8 +148,10 @@ if( NOT CXX_FLAGS_INITIALIZED )
       # Wiki page for more details
       # (https://gcc.gnu.org/wiki/Intel%20MPX%20support%20in%20the%20GCC%20compiler)
 
-      # string( APPEND CMAKE_C_FLAGS_DEBUG
-      #         " -fsanitize=float-divide-by-zero -fcheck-ponter-bounds")
+      string( APPEND CMAKE_C_FLAGS_DEBUG
+        " -fsanitize=float-divide-by-zero -fcheck-ponter-bounds")
```
Comment (alexrlongne) [inline config/unix-g++.cmake]: This should be "-fcheck-pointer-bounds" on lines 152 and 154.

Comment (KineticTheory) [inline config/unix-g++.cmake]: Nice catch.  I'll push an update in a couple of minutes.

Comment (KineticTheory) [inline config/unix-g++.cmake]: Additionally, `-fcheck-pointer-bounds` isn't supported on any of our machine.  I'm going to rework the logic a bit and turn on `-fsanitize=float-cast-overflow` instead.

**Comment by KineticTheory** [comment]
@alexrlongne Changes complete and tests show pass.  Read for final review and merge.

**Pull request merged**